### PR TITLE
feat(structured data) `core` endpoints to CRU rows

### PR DESCRIPTION
**Pull request by fontanierh**
list, get and upsert DB rows

remaining  code to complete https://github.com/dust-tt/dust/issues/2210

### CODE REVIEW

File: core/bin/dust_api.rs [hunk 1]
```diff
@@ -2076,6 +2210,18 @@ fn main() {
             "/projects/:project_id/data_sources/:data_source_id/databases/:database_id/tables",
             get(databases_tables_list),
         )
+        .route(
+            "/projects/:project_id/data_sources/:data_source_id/databases/:database_id/tables/:table_id",
```
Comment (spolu) [inline core/bin/dust_api.rs:2214]: rows as per other PR but since batch indeed no :row_id

File: core/bin/dust_api.rs [hunk 2]
```diff
@@ -2076,6 +2210,18 @@ fn main() {
...
+            post(databases_rows_upsert),
+        )
+        .route(
+            "/projects/:project_id/data_sources/:data_source_id/databases/:database_id/tables/:table_id/rows/:row_id",
+            get(databases_rows_retrieve),
+        )
+        .route(
+            "/projects/:project_id/data_sources/:data_source_id/databases/:database_id/rows",
```
Comment (spolu) [inline core/bin/dust_api.rs:2222]: should be tables/table_id ?

File: core/bin/dust_api.rs [hunk 3]
```diff
@@ -1871,6 +1871,140 @@ async fn databases_tables_list(
...
+            }),
+        ),
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct DatabasesRowsListQuery {
+    offset: usize,
+    limit: usize,
+    table_id: Option<String>,
```
Comment (spolu) [inline core/bin/dust_api.rs:1966]: I think it's a bad design to fetch all at once. We never want to do something with 2 rows from different tables and can always parallelize

**Review by spolu — CHANGES_REQUESTED** [review]

File: core/bin/dust_api.rs (hunk 3 above)
Comment (fontanierh) [reply core/bin/dust_api.rs:1966]: > We never want to do something with 2 rows

We kind of do when we generate the schema from scratch though, no ?

**Comment by fontanierh** [comment]
as per slack, merging and addressing comment in https://github.com/dust-tt/dust/pull/2489

**Pull request merged**